### PR TITLE
Surface guest feedback

### PR DESCRIPTION
**Pull request by deliav04**
## Hey, I just made a Pull Request!

TL;DR
Feedback submitted for an entity by a user that isn't in the catalog will now appear for the user that submitted it and any owners of the entity.

More info:
Although many of our employees are registered in the catalog as users, we also have many that we don't wish to add for a variety of reasons but would still like to reap the benefits of Backstage. Including submitting feedback on entities.
Before, any feedback submitted by a user that wasn't in the catalog would appear in the aggregates for the entity but to the user that submitted it it would appear as though their feedback wasn't recorded. The response also would not appear in the feedback section for the entity. 
These changes ensure that feedback is always visible to the user that submitted it to make it clear it was successfully received. And makes all feedback for an entity, including feedback submitted by users not in the catalog, visible to owners of the entity to surface feedback to those that need to see it.

This will also help address issues such as [this](https://github.com/backstage/backstage/issues/19340).

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/entity-feedback-backend/src/service/router.ts [hunk 1]
```diff
@@ -151,12 +151,23 @@ export async function createRouter(
...
 
     const token = getBearerTokenFromAuthorizationHeader(
       req.header('authorization'),
     );
 
+    const entity = (
+      await catalogClient.getEntitiesByRefs(
+        {
+          entityRefs: [req.params.entityRef],
+          fields: ['spec.ownerRef'],
```
Comment (kuangp) [inline plugins/entity-feedback-backend/src/service/router.ts:165]: I don't believe `spec.ownerRef` is a valid field - I assume you want `spec.owner`? If so, `spec.owner` is also not a standard field across all entities eg. User and Group entities don't have owners and even though rating users and team maybe not be a typical use case, this plugin current does allow for doing so

File: plugins/entity-feedback-backend/src/service/router.ts [hunk 2]
```diff
@@ -170,7 +181,16 @@ export async function createRouter(
...
 
-    res.json(ratings.filter(r => accessibleEntityRefs.includes(r.userRef)));
+    res.json(
+      ratings.filter(
+        r =>
+          accessibleEntityRefs.includes(r.userRef) ||
+          user?.identity.userEntityRef === r.userRef ||
+          user?.identity.ownershipEntityRefs.includes(
+            entity?.spec!.ownerRef as string,
+          ),
```
Comment (kuangp) [inline plugins/entity-feedback-backend/src/service/router.ts:191]: This can and should already be handled via an appropriate [permission policy](https://backstage.io/docs/permissions/writing-a-policy)

**Review by kuangp — CHANGES_REQUESTED** [review]
I don't think this change will work as you expect (have you tested this?) - From my understanding guest feedback will not be recorded since there is no identity tied to them (assuming you're using the default [guest user identity](https://github.com/backstage/backstage/blob/b74f157839a2a3f6b6051b58c27c9c360300bc75/packages/core-components/src/layout/SignInPage/GuestUserIdentity.ts#L23)) so this would change will not surface them since they are not persisted. 

Fundamentally, I don't think this plugin makes sense to support guest users since there is no way to differentiate one guest users feedback from another. For example, if one guest rates an entity and another guest views the same entity and gives it a different rating then it would overwrite the previous users feedback. This would also only result in one overall rating since there is no actual user to count these as separate ratings.

**Comment by deliav04** [comment]
> I don't think this change will work as you expect (have you tested this?) - From my understanding guest feedback will not be recorded 

I have tested it and the feedback is recorded in the database and is included in calculations for the aggregate like and dislikes but immediately disappears from the user's view after submission.

> Fundamentally, I don't think this plugin makes sense to support guest users since there is no way to differentiate one guest users feedback from another.
We're not using the default guest user. We're using the information [here](https://backstage.io/docs/auth/identity-resolver#sign-in-without-users-in-the-catalog) to assign tokens for guest users that aren't in the catalog. They each get assigned their own userRef so they are distinguishable and the userRef remains the same if they log back out and log back in. 

Happy to clarify things further if necessary!

**Comment by deliav04** [comment]
> > We're not using the default guest user. We're using the information [here](https://backstage.io/docs/auth/identity-resolver#sign-in-without-users-in-the-catalog) to assign tokens for guest users that aren't in the catalog. They each get assigned their own userRef so they are distinguishable and the userRef remains the same if they log back out and log back in.
> 
> Ah ok, that definitely helps clarify things here - ~in that case I would think this should work correctly for you (by default)~ so perhaps we need to debug the issue here a bit more to find the appropriate solution.
> 
> Edit: I think I understand the issue here now - still not too sure about the appropriate solution however but I still feel the current proposed one in this change might not be the best since it bypasses permissioning

I think the core of the issue is that right now the plugin is inconsistent for users not in the catalog. The feedback is recorded in the database and taken into consideration in aggregates but doesn't appear in the UI anywhere else. From the user's perspective it appears as though their feedback wasn't recorded even though it was. Ideally it could be fixed so that feedback from users not in the catalog is surfaced like any other feedback but it not being recorded at all would probably be preferred to the inconsistency of the current behavior. 
We weren't aware of the permissions framework, which should make it easy for us to hide feedback buttons from the UI for guest users as a temporary fix

**Comment by kuangp** [comment]
> We weren't aware of the permissions framework, which should make it easy for us to hide feedback buttons from the UI for guest users as a temporary fix

Yeah this might be a good temp fix for now. 

The main difficulty here is that we rely on the catalog permissioning to determine which users the current user has access to to filter the feedback so that the current user can only see responses from users they are authorized to be aware of - in other words, if there are "private users" that are permissioned not to be publicly discoverable then we don't want their feedback to be exposed as well (since that would reveal their existence) unless you are privileged to know about them. The problem for the non-catalog registered "guest users" use-case is that we don't know if said users either don't exist at all or we just don't have the authorization to know about their existence 😕

Perhaps one possible solution is to implement support for anonymous feedback which would truly support the default Guest identity user as well as non-catalog registered "guest users" that saves ratings/responses with a `null` defined userRef and have these ratings/responses always exposed to everyone (a registered user could also be able to choose to submit anonymously as well). However this would not solve the issue of indicating the previous feedback of the guest user but that might be more of a UI fix (maybe a confirmation popup that says "thank you your anonymous feedback has been recorded"). This also would open up the possibility for users to spam duplicate anonymous feedback but that probably comes with the territory.

**Comment by deliav04** [comment]
> The main difficulty here is that we rely on the catalog permissioning to determine which users the current user has access to to filter the feedback so that the current user can only see responses from users they are authorized to be aware of - in other words, if there are "private users" that are permissioned not to be publicly discoverable then we don't want their feedback to be exposed as well (since that would reveal their existence) unless you are privileged to know about them. The problem for the non-catalog registered "guest users" use-case is that we don't know if said users either don't exist at all or we just don't have the authorization to know about their existence 😕

Ohh OK that makes more sense.

> Perhaps one possible solution is to implement support for anonymous feedback which would truly support the default Guest identity user as well as non-catalog registered "guest users" that saves ratings/responses with a `null` defined userRef and have these ratings/responses always exposed to everyone (a registered user could also be able to choose to submit anonymously as well). However this would not solve the issue of indicating the previous feedback of the guest user but that might be more of a UI fix (maybe a confirmation popup that says "thank you your anonymous feedback has been recorded"). This also would open up the possibility for users to spam duplicate anonymous feedback but that probably comes with the territory.

I like that! Could the feedback still be stored with the userRef? That would allow for the current user check because if the feedback was submitted by the current user they should have permission to view it even if they're not in the catalog. Any other feedback submitted by users not in the catalog could be exposed as anonymous feedback.

**Comment by kuangp** [comment]
> Could the feedback still be stored with the userRef? That would allow for the current user check because if the feedback was submitted by the current user they should have permission to view it even if they're not in the catalog.

This would allow the feedback author to view their feedback but will run into the same issue of no one else being able to view it (assuming the author is a non-registered user) which make their feedback essentially useless

**Comment by deliav04** [comment]
> This would allow the feedback author to view their feedback but will run into the same issue of no one else being able to view it (assuming the author is a non-registered user) which make their feedback essentially useless

Sorry, maybe I wasn't being clear. I meant to implement your solution of anonymous feedback so that it can be exposed to everyone in an anonymous way but to still store the userRef of the person that submitted it. That way the userRef can be used to show the feedback to the person that submitted it and the feedback can be anonymized when exposing it to everyone else.

**Comment by kuangp** [comment]
> I meant to implement your solution of anonymous feedback so that it can be exposed to everyone in an anonymous way but to still store the userRef of the person that submitted it.

hm that wouldn't truly be anonymous though ...

**Comment by deliav04** [comment]
> hm that wouldn't truly be anonymous though ...

Is anything truly anonymous anymore? 😅

**Comment by kuangp** [comment]
> The main difficulty here is that we rely on the catalog permissioning to determine which users the current user has access to to filter the feedback so that the current user can only see responses from users they are authorized to be aware of - in other words, if there are "private users" that are permissioned not to be publicly discoverable then we don't want their feedback to be exposed as well (since that would reveal their existence) unless you are privileged to know about them. The problem for the non-catalog registered "guest users" use-case is that we don't know if said users either don't exist at all or we just don't have the authorization to know about their existence 😕

@Rugvip I'm curious if you have any thoughts about this from a security perspective - this seems like a more general permissions framework issue that could come up in various other use-cases in other areas of Backstage when dealing with [non-registered users](https://backstage.io/docs/auth/identity-resolver/#sign-in-without-users-in-the-catalog) and utilizing the catalog to cross reference user entity refs that are safe to expose

**Comment by Rugvip** [comment]
Yep tbh I think this is a rabbit hole we don't wanna go down and it's likely best to leave as is, gonna let this remain stale
